core/src/bytecode/typecheck/mod.rs

```diff
@@ -112,6 +133,7 @@ pub type TypeEnv<'ast> = Environment<Ident, UnifType<'ast>>;
 /// A term environment defined as a mapping from identifiers to a tuple of a term and an
 /// environment (i.e. a closure). Used to compute contract equality.
 #[derive(PartialEq, Clone, Debug)]
+//TODO: we should just store `&'ast Ast<'ast>` references here, instead of an owned version
```

Comment:
Why? Isn't `Ast` pretty small?

Reply:
One answer is because we can :slightly_smiling_face: 

Storing a ref isn't much harder than the owned `Ast`. While `Ast` is significantly smaller than `Term`, it is bigger than `RichTerm` and significantly bigger than `&'ast Ast<'ast>`: contrary to the former, we don't put the content behind an indirection (`Rc<Term>` for `RichTerm`), so it stores both the position and the node inline, taking up 56 bytes of storage currently vs 8 for the pointer alone. The node is much smaller than before (than `Term`), but it's still a large enum. In some sense, the right equivalent for `RichTerm` or `SharedTerm` is `&'ast Ast<'ast>` rather than `Ast<'ast>`.